ouroboros-consensus/src/Ouroboros/Consensus/Util/CBOR.hs

```diff
@@ -76,3 +89,40 @@ initDecoderIO getChunk = do
     checkEmpty :: ByteString -> Maybe ByteString
     checkEmpty bs | BS.null bs = Nothing
                   | otherwise  = Just bs
+
+{-------------------------------------------------------------------------------
+  HasFS interaction
+-------------------------------------------------------------------------------}
+
+data ReadIncrementalErr =
+    -- | Could not deserialise the data
+    ReadFailed S.DeserialiseFailure
+
+    -- | Deserialisation was successful, but there was additional data
+  | TrailingBytes ByteString
+
+-- | Read a file incrementally
+readIncremental :: forall m h a. (Serialise a, MonadST m, MonadThrow m)
+                => HasFS m h -> FsPath -> m (Either ReadIncrementalErr a)
+readIncremental hasFS@HasFS{..} fp = withLiftST $ \f -> do
+    withFile hasFS fp ReadMode $ \h ->
+      go f h =<< f S.deserialiseIncremental
+  where
+    go :: (forall x. ST s x -> m x)
+       -> h
+       -> S.IDecode s a
+       -> m (Either ReadIncrementalErr a)
+    go f h (S.Partial k) = do
+        bs   <- hGet h defaultChunkSize
```

Comment:
Re #277 is this `hGetSome` or `hGetExactly`? I presume `hGetSome`, which is fine.

Reply:
Re `defaultChunkSize`, are we going to be using this for small reads, like single blocks, or for streaming bigly chunks? The `defaultChunkSize` is 32k or 64k iirc which is good for I/O performance if you're reading a lot.

Reply:
Primarily use case right now is reading the ledger state. I'll add a comment we might want to reconsider this if we use it for small examples.